substrate/zip_container

```diff
@@ -0,0 +1,366 @@
+			{
+				return get<I>() >= rhs.template get<I>();
+			}
+
+#if __cplusplus >= 201703L
+			template<size_t... Is>
+			inline bool ge(const zipIterator_t<Iterators...> &rhs, indexSequence_t<Is...>) const noexcept
+			{
+				return (ge<Is>(rhs) && ...);
+			}
+#else
+			template<size_t... Is>
+			inline bool ge(const zipIterator_t<Iterators...> &rhs, indexSequence_t<Is...>) const noexcept
+			{
+				const std::array<bool, length> equals = {{eq<Is...>(rhs)}};
+				return std::all_of(equals.begin(), equals.end(), [](const bool &value) { return value; });
+			}
+#endif
+
+			template<size_t I> inline bool eq(const zipIterator_t<Iterators...> &rhs) const noexcept
+			{
+				return rhs.template get<I>() == get<I>();
+			}
+
+#if __cplusplus >= 201703L
+			template<size_t... Is>
+			inline bool eq(const zipIterator_t<Iterators...> &rhs, indexSequence_t<Is...>) const noexcept
+			{
+				return (eq<Is>(rhs) && ...);
+			}
+#else
+			template<size_t... Is>
+			inline bool eq(const zipIterator_t<Iterators...> &rhs, indexSequence_t<Is...>) const noexcept
+			{
+				const std::array<bool, length> equals = {{eq<Is>(rhs)...}};
+				return std::all_of(equals.begin(), equals.end(), [](const bool &value) { return value; });
+			}
+#endif
+
+			template<size_t... Is> inline reference getAll(indexSequence_t<Is...>) noexcept
+			{
+#if __cplusplus >= 201703L
+				return {(get<Is>())...};
+#else
+				return {get<Is>()...};
+#endif
+			}
+
+			template<size_t... Is> inline const_reference getAll(indexSequence_t<Is...>) const noexcept
+			{
+#if __cplusplus >= 201703L
+				return {get<Is...>()};
+#else
+				return {get<Is>()...};
+#endif
+			}
+
+		public:
+			// NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+			zipIterator_t(Iterators &&...its, size_t index, size_t containerLength) :
+				index(index), containerLength(containerLength), iterators(std::forward<Iterators>(its)...)
+			{
+				if (containerLength > static_cast<size_t>(std::numeric_limits<difference_type>::max()))
+				{
+					throw std::out_of_range("Accessing iterators beyond PTRDIFF_MAX is undefined behaviour");
+				}
+
+				if (index > containerLength)
+				{
+					throw std::out_of_range("Accessing iterators beyond boundaries is undefined behaviour");
+				}
+			}
+
+			zipIterator_t(zipIterator_t &&) noexcept = default;
+			zipIterator_t &operator=(zipIterator_t &&) noexcept = default;
+
+			~zipIterator_t() noexcept = default;
+
+			SUBSTRATE_NO_DISCARD(inline zipIterator_t operator+(const size_t value) const noexcept)
+			{
+				zipIterator_t<Iterators...> result{*this};
+				result += value;
+				return result;
+			}
+
+			SUBSTRATE_NO_DISCARD(inline zipIterator_t operator-(const size_t value) const noexcept)
+			{
+				zipIterator_t<Iterators...> result{*this};
+				result -= value;
+				return result;
+			}
+
+			inline zipIterator_t &operator+=(const size_t value)
+			{
+				if (index > std::numeric_limits<size_t>::max() - value)
+				{
```

Comment:
Same comment about braces (we'll make this the last so we're not repeating ourself too much but it applies down the whole file)